File: fli-js/src/core/currency.ts
/**
 * Currency extraction from Google Flights price tokens.
 *
 * 1:1 port of fli/core/currency.py — varint walk over a base64-urlsafe
 * protobuf payload to extract the nested ISO currency code, cached to
 * avoid repeated decoding for the common one-currency-per-response case.
 */

const _decodeCache = new Map<string, string | null>();
const MAX_CACHE = 256;

function readVarint(data: Uint8Array, offset: number): [number, number] {
  // `+= chunk * 2**shift` instead of `|= chunk << shift` — JavaScript's
  // bitwise operators coerce to signed 32-bit and would corrupt the
  // result for shifts ≥ 31 (i.e. varint values ≥ 2^31).
  let value = 0;
  let shift = 0;
  let off = offset;
  while (off < data.length) {
    const byte = data[off];
    if (byte === undefined) break;
    off += 1;
    value += (byte & 0x7f) * 2 ** shift;
    if ((byte & 0x80) === 0) {
      if (!Number.isSafeInteger(value)) {
        throw new Error("Varint exceeds Number.MAX_SAFE_INTEGER");
      }
      return [value, off];
    }
    shift += 7;
    if (shift >= 64) throw new Error("Varint is too large to decode");
  }
  throw new Error("Unexpected end of data while decoding varint");
}

function readLengthDelimited(data: Uint8Array, offset: number): [Uint8Array, number] {
  const [length, after] = readVarint(data, offset);
  const end = after + length;
  if (end > data.length) throw new Error("Length-delimited field exceeds payload size");
  return [data.subarray(after, end), end];
}

function skipField(data: Uint8Array, offset: number, wireType: number): number {
  if (wireType === 0) return readVarint(data, offset)[1];
  if (wireType === 1) {
    const end = offset + 8;
    if (end > data.length) throw new Error("Fixed64 field exceeds payload size");
    return end;
  }
  if (wireType === 2) return readLengthDelimited(data, offset)[1];
  if (wireType === 5) {
    const end = offset + 4;
    if (end > data.length) throw new Error("Fixed32 field exceeds payload size");
    return end;
  }
  throw new Error(`Unsupported wire type: ${wireType}`);
}

function extractCurrencyFromMessage(data: Uint8Array): string | null {
  let offset = 0;
  while (offset < data.length) {
    const [tag, afterTag] = readVarint(data, offset);
    offset = afterTag;
    const fieldNumber = tag >> 3;
    const wireType = tag & 0x07;

    if (fieldNumber === 3 && wireType === 2) {
      const [nested, after] = readLengthDelimited(data, offset);
      offset = after;
      let nestedOffset = 0;
      while (nestedOffset < nested.length) {
        const [ntag, na] = readVarint(nested, nestedOffset);
        nestedOffset = na;
        const nfield = ntag >> 3;
        const nwire = ntag & 0x07;
        if (nfield === 3 && nwire === 2) {
          const [bytes, _end] = readLengthDelimited(nested, nestedOffset);
          return new TextDecoder("utf-8").decode(bytes).toUpperCase();
        }
        nestedOffset = skipField(nested, nestedOffset, nwire);
      }
      continue;
    }

    offset = skipField(data, offset, wireType);
  }
  return null;
}

function base64UrlsafeDecode(token: string): Uint8Array {
  const padLen = (4 - (token.length % 4)) % 4;
  const padded = token + "=".repeat(padLen);
  const standard = padded.replace(/-/g, "+").replace(/_/g, "/");
  if (typeof atob !== "undefined") {
    const binary = atob(standard);
    const out = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
    return out;
  }
  return new Uint8Array(Buffer.from(standard, "base64"));
}

/**
 * Extract the ISO currency code from a Google Flights price token.
 *
 * Cached so the same token returned for every row in a response decodes
 * the protobuf payload exactly once.
 */
export function extractCurrencyFromPriceToken(token: string | null | undefined): string | null {
  if (!token) return null;
  if (_decodeCache.has(token)) return _decodeCache.get(token) ?? null;
  let result: string | null = null;
  try {
    const decoded = base64UrlsafeDecode(token);
    result = extractCurrencyFromMessage(decoded);
  } catch {
    result = null;
  }
  // LRU-ish bound: drop oldest if we hit the cap.
  if (_decodeCache.size >= MAX_CACHE) {
    const firstKey = _decodeCache.keys().next().value;
    if (firstKey !== undefined) _decodeCache.delete(firstKey);
  }
  _decodeCache.set(token, result);
  return result;
}

/** Currency-code-aware formatter built on `Intl.NumberFormat`. */
export function formatPrice(
  amount: number | null | undefined,
  currencyCode: string | null | undefined,
): string {
  if (amount == null) {
    return currencyCode ? `${currencyCode.toUpperCase()} —` : "—";
  }
  if (!currencyCode) {
    return amount.toLocaleString("en-US", {
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    });
  }
  const normalized = currencyCode.toUpperCase();
  try {
    return new Intl.NumberFormat("en-US", {
      style: "currency",
      currency: normalized,
    }).format(amount);
  } catch {
    return `${normalized} ${amount.toLocaleString("en-US", {
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    })}`;
  }
}

/** Build a chart-axis label for one or more result currencies. */
export function formatPriceAxisLabel(currencies: Iterable<string | null | undefined>): string {
  const set = new Set<string>();
  for (const c of currencies) {
    if (c) set.add(c.toUpperCase());
  }
  if (set.size === 1) {
    const [only] = set;
    return `Price (${only})`;
  }
  return "Price";
}

/** Clear the in-memory token-decode cache (test helper). */
export function _clearCurrencyCache(): void {
  _decodeCache.clear();
}

File: fli/core/currency.py
"""Utilities for extracting and formatting price currencies."""

from __future__ import annotations

import base64
from collections.abc import Iterable
from functools import lru_cache

from babel.numbers import format_currency as babel_format_currency


def _read_varint(data: bytes, offset: int) -> tuple[int, int]:
    """Read a protobuf-style varint value."""
    value = 0
    shift = 0

    while offset < len(data):
        byte = data[offset]
        offset += 1
        value |= (byte & 0x7F) << shift
        if not byte & 0x80:
            return value, offset

        shift += 7
        if shift >= 64:
            raise ValueError("Varint is too large to decode")

    raise ValueError("Unexpected end of data while decoding varint")


def _read_length_delimited(data: bytes, offset: int) -> tuple[bytes, int]:
    """Read a protobuf-style length-delimited field."""
    length, offset = _read_varint(data, offset)
    end = offset + length
    if end > len(data):
        raise ValueError("Length-delimited field exceeds payload size")
    return data[offset:end], end


def _skip_field(data: bytes, offset: int, wire_type: int) -> int:
    """Skip over a protobuf field we do not need."""
    if wire_type == 0:
        _, offset = _read_varint(data, offset)
        return offset
    if wire_type == 1:
        end = offset + 8
        if end > len(data):
            raise ValueError("Fixed64 field exceeds payload size")
        return end
    if wire_type == 2:
        _, offset = _read_length_delimited(data, offset)
        return offset
    if wire_type == 5:
        end = offset + 4
        if end > len(data):
            raise ValueError("Fixed32 field exceeds payload size")
        return end
    raise ValueError(f"Unsupported wire type: {wire_type}")


def _extract_currency_from_message(data: bytes) -> str | None:
    """Extract the nested ISO currency code from a decoded token."""
    offset = 0

    while offset < len(data):
        tag, offset = _read_varint(data, offset)
        field_number = tag >> 3
        wire_type = tag & 0x07

        if field_number == 3 and wire_type == 2:
            nested_message, offset = _read_length_delimited(data, offset)
            nested_offset = 0
            while nested_offset < len(nested_message):
                nested_tag, nested_offset = _read_varint(nested_message, nested_offset)
                nested_field = nested_tag >> 3
                nested_wire_type = nested_tag & 0x07

                if nested_field == 3 and nested_wire_type == 2:
                    currency_bytes, nested_offset = _read_length_delimited(
                        nested_message, nested_offset
                    )
                    return currency_bytes.decode("utf-8").upper()

                nested_offset = _skip_field(nested_message, nested_offset, nested_wire_type)
            continue

        offset = _skip_field(data, offset, wire_type)

    return None


def extract_currency_from_price_token(token: str | None) -> str | None:
    """Extract the ISO currency code from a Google Flights price token.

    Cached so the same token returned for every row in a response (which
    is the common case — one currency per response) decodes the protobuf
    payload exactly once. The varint walk is ~2.5us per call cold; cache
    hits are ~100ns, a ~25x speedup on the parsing hot path.
    """
    return _decode_token(token) if token else None


@lru_cache(maxsize=256)
def _decode_token(token: str) -> str | None:
    """Pure-function inner — guarded for None so the cache only sees ``str``."""
    try:
        padded_token = token + ("=" * (-len(token) % 4))
        decoded = base64.urlsafe_b64decode(padded_token)
        return _extract_currency_from_message(decoded)
    except (UnicodeDecodeError, ValueError, base64.binascii.Error):
        return None


def format_price(amount: float | None, currency_code: str | None) -> str:
    """Format a price using its ISO currency code.

    ``amount`` may be ``None`` when Google did not surface a per-row price
    (premium-cabin round-trips often hit this — see
    :class:`fli.models.FlightResult`). The placeholder ``"—"`` (em dash)
    is returned in that case, matching the convention used elsewhere in
    the CLI display for unknown structured fields.
    """
    if amount is None:
        return f"{currency_code.upper()} —" if currency_code else "—"
    if not currency_code:
        return f"{amount:,.2f}"

    normalized_currency = currency_code.upper()
    try:
        return babel_format_currency(amount, normalized_currency, locale="en_US")
    except (TypeError, ValueError):
        return f"{normalized_currency} {amount:,.2f}"


def format_price_axis_label(currencies: Iterable[str | None]) -> str:
    """Build a chart axis label for one or more result currencies."""
    normalized = {currency.upper() for currency in currencies if currency}
    if len(normalized) == 1:
        return f"Price ({normalized.pop()})"
    return "Price"
